### PR TITLE
Bluetooth: Audio: Fixing discovery and subscription

### LINKED ISSUE #67434: Disconnection might cause the unexpected behavior on unicast client
**Describe the bug**
When using nRF5340 audio (which is using Zephyr as host) for setting up CIS TWS demo, if we reset both unicast server at the same time, there could be a chance the unicast client just stuck.

**To Reproduce**
To reproduce with NCS and nRF5340 audio DK
1. Setup the nrf5340 audio TWS demo
2. Once if gateway connects to both L and R headsets, press pause button for pausing the audio stream
3. Press reset button on both L and R headsets at the same time
4. After multiple trials on steps 3, there could be a chance the gateway just stuck

**Logs and console output**
This is the Ozone debugger scan capture when the issue triggered:
ozone_screen_capture

And we can see the code is trapped in following parts
trapped_code

After enabing
CONFIG_BT_GATT_LOG_LEVEL_DBG=y
CONFIG_BT_BAP_UNICAST_CLIENT_LOG_LEVEL_DBG=y

The disconnection part of log before issue triggered:
`
GW [00:00:44.198,425] <dbg> bt_bap_unicast_client: unicast_client_ep_qos_state: dir sink cig 0x00 cis 0x00 codec 0x06 interval 10000 framing 0x00 phy 0x02 rtn 2 latency 20 pd 10000

GW [00:00:53.817,962] <dbg> bt_gatt: bt_gatt_disconnected: conn 0x20008900

GW [00:00:53.818,145] <dbg> bt_bap_unicast_client: unicast_client_cp_notify: conn 0x20008900 len 0
GW [00:00:53.818,176] <dbg> bt_bap_unicast_client: unicast_client_cp_notify: Unsubscribed
GW [00:00:53.818,176] <wrn> bt_att: Not connected
GW [00:00:53.818,176] <dbg> bt_bap_unicast_client: unicast_client_ep_notify: conn 0x20008900 ep 0x20009040 len 0
GW [00:00:53.818,206] <dbg> bt_bap_unicast_client: unicast_client_ep_notify: Unsubscribed

GW [00:00:53.818,206] <dbg> bt_bap_unicast_client: unicast_client_disconnected: conn 0x20008900 reason 0x08

GW [00:00:53.818,206] <dbg> bt_bap_unicast_client: unicast_client_ep_reset: conn 0x20008900
GW [00:00:53.818,237] <dbg> bt_bap_unicast_client: unicast_client_reset: ep 0x20009040
GW [00:00:53.818,267] <dbg> bt_bap_unicast_client: unicast_client_ep_set_status: ep 0x20009040 handle 0x0013 id 0x01 dir sink state qos-configured -> idle
GW [00:00:53.818,359] <dbg> bt_bap_unicast_client: unicast_client_reset: ep 0x20009128
GW [00:00:53.818,389] <dbg> bt_bap_unicast_client: unicast_client_ep_set_status: ep 0x20009128 handle 0x0000 id 0x00 dir Unknown state idle -> idle
GW [00:00:53.818,481] <inf> bt_mgmt: Disconnected: FD:55:BF:84:E7:97 (random) (reason 0x08)
GW [00:00:53.818,511] <inf> streamctrl_unicast_client: Device disconnected
GW [00:00:53.818,664] <inf> bt_mgmt_scan: All bonded slots filled, will not accept new devices
GW [00:00:53.831,268] <inf> bt_mgmt_scan: Local addr: 78:27:70:FC:65:1E (random). May time out. Updates not printed
GW [00:00:53.831,298] <inf> bt_mgmt_scan: Scanning successfully started

GW [00:00:53.837,951] <dbg> bt_gatt: bt_gatt_disconnected: conn 0x200089c8

GW [00:00:53.837,982] <dbg> bt_gatt: disconnected_cb: ccc 0x20000f88 reseted
GW [00:00:53.837,982] <dbg> bt_gatt: disconnected_cb: ccc 0x20000f2c reseted
GW [00:00:53.838,012] <dbg> bt_gatt: disconnected_cb: ccc 0x20000ee4 reseted
GW [00:00:53.838,012] <dbg> bt_gatt: disconnected_cb: ccc 0x20000e9c reseted
GW [00:00:53.838,043t: disconnected_cb: ccc 0x20000e54 reseted
GW [00:00:53.838,043] <dbg> bt_gatt: disconnected_cb: ccc 0x20000e0c reseted
GW [00:00:53.838,073] <dbg> bt_gatt: disconnected_cb: ccc 0x20000dc4 reseted
GW [00:00:53.838,073] <dbg> bt_gatt: disconnected_cb: ccc 0x20000d7c reseted
GW [00:00:53.838,073] <dbg> bt_gatt: disconnected_cb: ccc 0x20000d14 reseted
GW [00:00:53.838,104] <dbg> bt_gatt: disconnected_cb: ccc 0x20000ccc reseted
GW [00:00:53.838,104] <dbg> bt_gatt: disconnected_cb: ccc 0x20000c84 reseted
GW [00:00:53.838,195] <dbg> bt_bap_unicast_client: unicast_client_cp_notify: conn 0x200089c8 len 0
GW [00:00:53.838,195] <dbg> bt_bap_unicast_client: unicast_client_cp_notify: Unsubscribed
GW [00:00:53.838,195] <wrn> bt_att: Not connected
GW [00:00:53.838,226] <dbg> bt_bap_unicast_client: unicast_client_ep_notify: conn 0x200089c8 ep 0x200094f8 len 0
GW [00:00:53.838,226] <dbg> bt_bap_unicast_client: unicast_client_ep_notify: Unsubscribed

GW [00:00:53.838,226] <dbg> bt_bap_unicast_client: unicast_client_disconnected: conn 0x200089c8 reason 0x08
`

### PR DESCRIPTION
**Pull request by Thalley**
The way that bt_gatt_subscribe is used by the LE Audio GATT clients were subpar. They were built on the assumption that we are not using bonding. None of the clients really support proper bonding, and thus the subscriptions should not be considered working after a disconnection. 

For this `BT_GATT_SUBSCRIBE_FLAG_VOLATILE` has been used, until we support bonding. 

The discovery functions have been modified to better support this as well, except for the broadcast assistant and media control client that have some other fundamental issue with multiple connections. 

One BSIM test was disabled as that expected bonding support on the client. 
Multiple BSIM tests were updated to verify that we can at least do multiple discovery calls for a single connection, but should be updated to also verify that discovery and other features work after a reconnect, but that is a TODO for now. 

fixes https://github.com/zephyrproject-rtos/zephyr/issues/67434

### CODE REVIEW

File: subsys/bluetooth/audio/aics_client.c [hunk 1]
```diff
@@ -616,7 +618,14 @@ static uint8_t aics_discover_func(struct bt_conn *conn, const struct bt_gatt_att
 			 */
 			sub_params->ccc_handle = attr->handle + 2;
 			sub_params->notify = aics_client_notify_handler;
-			bt_gatt_subscribe(conn, sub_params);
+			err = bt_gatt_subscribe(conn, sub_params);
+			if (err != 0 && err != -EALREADY) {
+				LOG_ERR("Failed to subscribe: %d", err);
+
+				inst->cli.cb->discover(inst, err);
```
Comment (fredrikdanebjer) [inline subsys/bluetooth/audio/aics_client.c:627]: Shouldn't this be guarded, as is done further up in the function before usage too?
`if (inst->cli.cb && inst->cli.cb->discover) {`

File: subsys/bluetooth/audio/bap_broadcast_assistant.c [hunk 2]
```diff
@@ -594,6 +595,63 @@ static struct bt_le_scan_cb scan_cb = {
...
+	inst->recv_state_cnt = 0U;
+	inst->start_handle = 0U;
+	inst->end_handle = 0U;
+	inst->cp_handle = 0U;
+
+	for (int i = 0U; i < CONFIG_BT_BAP_BROADCAST_ASSISTANT_RECV_STATE_COUNT; i++) {
+		inst->src_ids[i] = 0U;
+		inst->past_avail[i] = false;
+		inst->recv_state_handles[i] = 0U;
+		inst->src_ids[i] = 0U;
```
Comment (fredrikdanebjer) [inline subsys/bluetooth/audio/bap_broadcast_assistant.c:612]: This is a duplication from 3 lines up

File: subsys/bluetooth/audio/bap_broadcast_assistant.c [hunk 3]
```diff
@@ -594,6 +595,63 @@ static struct bt_le_scan_cb scan_cb = {
...
+					return err;
+				}
+			}
+		}
+
+		bt_conn_unref(conn);
+		inst->conn = NULL;
+	}
+
+	return 0;
```
Comment (fredrikdanebjer) [inline subsys/bluetooth/audio/bap_broadcast_assistant.c:643]: There are some members of inst that we dont seem to reset, do we not need to do that?
recv_state_disc_params
read_params
write_params
disc_params

**Review by fredrikdanebjer — COMMENTED** [review]
In the commit message of `Bluetooth: Audio: Use BT_GATT_SUBSCRIBE_FLAG_VOLATILE`, in the third paragraph, this doesn't seem right:
`they are have no`

**Review by fredrikdanebjer — DISMISSED** [review]

**Comment by Thalley** [comment]
Rebased so solve merge conflicts

**Review by fredrikdanebjer — DISMISSED** [review]

**Review by kruithofa — DISMISSED** [review]

**Review by fredrikdanebjer — DISMISSED** [review]

**Comment by carlescufi** [comment]
@Thalley please rebase

**Comment by Thalley** [comment]
Rebased to solve merge conflicts

**Review by fredrikdanebjer — APPROVED** [review]

**Review by kruithofa — APPROVED** [review]

**Pull request merged**